util/exponential-backoff.js

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+const DEFAULT_MAX_RETRIES = 5
+const DEFAULY_RETRY_MULTIPLIER_MS = 100
+const DEFAULT_RETRY_COUNT = 0
+
+const logger = require('@adobe/aio-lib-core-logging')('@adobe/aio-lib-core-tvm', { provider: 'debug' })
+const fetch = require('node-fetch').default
+const util = require('util')
+const sleep = util.promisify(setTimeout)
+
+/**
+ * This function will retry connecting to a url end-point, with
+exponential backoff. Returns a Promise.
+ *
+ * @param {string} url endpoint url
+ * @param {object} requestOptions request options
+ * @param {object} retryOptions retry options with keys being maxRetries, retryMultipler and retryCount
+ * @returns {Promise} Promise object representing the http response
+ */
+async function exponentialBackoff (url, requestOptions, retryOptions) {
+  const {
+    maxRetries = DEFAULT_MAX_RETRIES,
+    retryMultiplier = DEFAULY_RETRY_MULTIPLIER_MS,
+    retryCount = DEFAULT_RETRY_COUNT
+  } = retryOptions || {}
+  return _exponentialBackoffHelper(url, requestOptions, { maxRetries, retryMultiplier, retryCount })
+}
+
+/**
+ * Helper function
+ *
+ * @param {string} url endpoint url
+ * @param {object} requestOptions request options
+ * @param {number} retryOptions retry options with keys being maxRetries, retryMultipler and retryCount
+ * @returns {object} http response
+ * @private
+ */
+async function _exponentialBackoffHelper (url, requestOptions, retryOptions) {
+  const { maxRetries, retryMultiplier, retryCount } = retryOptions
+
+  const response = await fetch(url, requestOptions)
```

Comment:
I think we should also cover the case of connection errors i.e. when fetch throws

like `try { await fetch } catch (e) { retry if retry count is < max }`

Reply:
@moritzraho Ya i have seen it but I m not so sure about this. It's essentially a client error. For eg: fetch throws when the domain in the url does not exist. For such errors, nothing changes in a few seconds. I think retry makes sense mainly in case of server errors.

Reply:
Well I think it's safer to consider also network errors it could be a temporary connection or DNS failure